### PR TITLE
feat(mesh): add story more action

### PR DESCRIPTION
**Pull request by caesarWHLee**
# Feature
Add story more action button with following feature:
- redirect to publisher's donate page
- add/remove bookmark
- unfollow publisher (media page only)
- copy story link to user clipboard
- share story link to media
  - Facebook
  - Line
  - Threads
  - X

## The position of `ActionSheet` and the scroll-to-close logic
The `ActionSheet`'s default CSS is `position: absolute` to easily stick to the button, but in some scenario this won't work.
圖片

Like the screenshot above, the absolute action sheet will be inside the scroll container which is not what we want. So for this situation we need to switch to `position: fixed` and set the dynamic `top` and `left` according to the button position.

As the user scrolls the button's position will change and we need to update the new position for the action sheet which can be achieved by leveraging listening to scroll event and keep setting the new position which is not good for the performance. 

So here we simply close the action sheet to avoid the complicated logic. This approach will apply to other scenario even it may not need this to align the behavior.

## Follow-up
- Add toast component to show copied successfully

### CODE REVIEW

**Comment by caesarWHLee** [comment]
@LinHeMa @tyc67 @nihclil 一樣請大家看一下各自頁面的 moreAction 改動有沒有問題，謝啦

File: packages/mesh/app/actions/follow-publisher.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  memberId,
+  publisherId,
+}: {
+  memberId: string
+  publisherId: string
+}) {
+  const payload = {
+    action: 'add_bookmark',
+    memberId,
+    objective: 'publisher',
```
Comment (tyc67) [inline packages/mesh/app/actions/follow-publisher.ts:16]: 提問: Ln.16, 38 應該用 `objective: publisher` 還是 `objective: member` ?

File: packages/mesh/components/story-more-action-button.tsx [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+    onClose,
+  }: {
+    storyId: string
+    onClose: () => void
+  },
+  ref: ForwardedRef<HTMLDivElement>
+) {
+  const getShareUrl = (template: string) => {
+    const storyUrl = getStoryUrl(storyId)
+    return template.replace('${storyUrl}', encodeURIComponent(storyUrl))
```
Comment (tyc67) [inline packages/mesh/components/story-more-action-button.tsx]: 提問: story url 是 a-z,0-9 且不含特殊字元, 是否需要 encode ? 如果要 encode url 是否使用 [encodeURI( )](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURI)

File: packages/mesh/components/story-more-action-button.tsx [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+    <div
+      className="fixed inset-0 z-modal flex  items-center justify-center  bg-lightbox-light"
+      ref={ref}
+    >
+      <div className="w-[335px] rounded-xl bg-white shadow-light-box sm:w-[480px]">
+        <div className="flex h-15 items-center justify-between border-b border-[rgba(0,9,40,0.1)] px-2">
+          <div />
+          <div className="list-title text-primary-800">分享</div>
+          <button
+            className="size11 flex items-center justify-center"
```
Comment (tyc67) [inline packages/mesh/components/story-more-action-button.tsx]: size-11 ?

File: packages/mesh/components/story-more-action-button.tsx [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+  storyId: string
+  publisherId: string
+  canUnFollowPublisher?: boolean
+  showOnRestrictArea?: boolean
+  nestedScrollContainerRef?: RefObject<HTMLElement>
+  className?: string
+}) {
+  const [shouldShowShareSheet, setShouldShowShareSheet] = useState(false)
+  const [shouldShowActionSheet, setShouldShowActionSheet] = useState(false)
+  const [position, setPosition] = useState<Position>({ top: NaN, left: NaN })
```
Comment (tyc67) [inline packages/mesh/components/story-more-action-button.tsx]: 初始值可能用 `null` 或 `Infinity` 比較好

File: packages/mesh/components/story-more-action-button.tsx [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+    </div>
+  )
+})
+
+const shareMedia = [
+  {
+    id: 'facebook',
+    icon: 'icon-share-facebook',
+    text: 'Facebook',
+    template: 'https://www.facebook.com/share.php?u=${storyUrl}',
```
Comment (tyc67) [inline packages/mesh/components/story-more-action-button.tsx]: 建議可以 naming 為 `templateUrl` 增加可讀性和搜尋

File: packages/mesh/components/story-more-action-button.tsx [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+
+const isPositionValid = (position: Position) => {
+  return Number.isFinite(position.top) && Number.isFinite(position.left)
+}
+
+const getStoryUrl = (storyId: string) => {
+  return `${location.origin}/story/${storyId}`
+}
+
+export default function StoryMoreActionButton({
```
Comment (tyc67) [inline packages/mesh/components/story-more-action-button.tsx:29]: 是否需要 handle `storyId` , `publisherId` props value 為 empty string 的情況 ?

**Comment by nihclil** [comment]
截圖 2024-10-08 中午12 48 03

shresheet clickOutside 好像無法成功～？

**Comment by caesarWHLee** [comment]
> 截圖 2024-10-08 中午12 48 03
> 
> shresheet clickOutside 好像無法成功～？

這邊寫錯，改成使用點擊背景關閉，修改在
[559f6e0](https://github.com/readr-media/Sachiel/pull/861/commits/559f6e0d3ef3fecf62dfa03cc7aeb9509efe1992)

File: packages/mesh/app/actions/follow-publisher.ts (hunk 1 above)
Comment (caesarWHLee) [reply packages/mesh/app/actions/follow-publisher.ts:16]: `objective: member` > follow member
`objective: publisher` > follow publisher

倒是上面的 action 沒改到，修正為 'add_follow'，修改在 [cd42cdb](https://github.com/readr-media/Sachiel/pull/861/commits/cd42cdb558108ca892dfa38b614b06466602d6a4)

File: packages/mesh/components/story-more-action-button.tsx (hunk 2 above)
Comment (caesarWHLee) [reply packages/mesh/components/story-more-action-button.tsx]: 拿掉 `encodeURIComponent`，感謝提醒，修改在[d91c124](https://github.com/readr-media/Sachiel/pull/861/commits/d91c124873de1acd3fa83fce42d54a79d1a3c13a)

File: packages/mesh/components/story-more-action-button.tsx (hunk 3 above)
Comment (caesarWHLee) [reply packages/mesh/components/story-more-action-button.tsx]: 修改在 [559f6e0](https://github.com/readr-media/Sachiel/pull/861/commits/559f6e0d3ef3fecf62dfa03cc7aeb9509efe1992)

File: packages/mesh/components/story-more-action-button.tsx (hunk 4 above)
Comment (caesarWHLee) [reply packages/mesh/components/story-more-action-button.tsx]: 改成 Infinity，修改在[d91c124](https://github.com/readr-media/Sachiel/pull/861/commits/d91c124873de1acd3fa83fce42d54a79d1a3c13a)

File: packages/mesh/components/story-more-action-button.tsx (hunk 5 above)
Comment (caesarWHLee) [reply packages/mesh/components/story-more-action-button.tsx]: 修改在 [d91c124](https://github.com/readr-media/Sachiel/pull/861/commits/d91c124873de1acd3fa83fce42d54a79d1a3c13a)

File: packages/mesh/components/story-more-action-button.tsx (hunk 6 above)
Comment (caesarWHLee) [reply packages/mesh/components/story-more-action-button.tsx:29]: 這邊處理，[fd25f81](https://github.com/readr-media/Sachiel/pull/861/commits/fd25f81cf473e73fd34ab9667edc2c347cdd2aa4)，等 toast 完成後顯示錯誤訊息

**Review by tyc67 — APPROVED** [review]

**Comment by caesarWHLee** [comment]
突然想到 @LinHeMa 負責的 profile 頁面沒有用到這次的 `<StoryMoreActionButton/>` 先拿掉 reviewer

**Pull request merged**